File: ja/codes/swift/chapter_divide_and_conquer/binary_search_recur.swift
/**
 * File: binary_search_recur.swift
 * Created Time: 2023-09-02
 * Author: nuomi1 (nuomi1@qq.com)
 */

/* 二分探索：問題 f(i, j) */
func dfs(nums: [Int], target: Int, i: Int, j: Int) -> Int {
    // 区間が空なら対象要素は存在しないので -1 を返す
    if i > j {
        return -1
    }
    // 中点インデックス m を計算
    let m = (i + j) / 2
    if nums[m] < target {
        // 部分問題 f(m+1, j) を再帰的に解く
        return dfs(nums: nums, target: target, i: m + 1, j: j)
    } else if nums[m] > target {
        // 部分問題 f(i, m-1) を再帰的に解く
        return dfs(nums: nums, target: target, i: i, j: m - 1)
    } else {
        // 目標要素が見つかったらそのインデックスを返す
        return m
    }
}

/* 二分探索 */
func binarySearch(nums: [Int], target: Int) -> Int {
    // 問題 f(0, n-1) を解く
    dfs(nums: nums, target: target, i: nums.startIndex, j: nums.endIndex - 1)
}

@main
enum BinarySearchRecur {
    /* Driver Code */
    static func main() {
        let target = 6
        let nums = [1, 3, 6, 8, 12, 15, 23, 26, 31, 35]

        // 二分探索（両閉区間）
        let index = binarySearch(nums: nums, target: target)
        print("対象要素 6 のインデックス = \(index)")
    }
}

File: vi/codes/go/chapter_divide_and_conquer/binary_search_recur.go
// File: binary_search_recur.go
// Created Time: 2023-07-19
// Author: hongyun-robot (1836017030@qq.com)

package chapter_divide_and_conquer

/* Binary search: problem f(i, j) */
func dfs(nums []int, target, i, j int) int {
	// If interval is empty, indicating no target element, return -1
	if i > j {
		return -1
	}
	// Calculate midpoint index
	m := i + ((j - i) >> 1)
	// Compare midpoint with target element
	if nums[m] < target {
		// If smaller, recurse on right half of array
		// Recursion subproblem f(m+1, j)
		return dfs(nums, target, m+1, j)
	} else if nums[m] > target {
		// If larger, recurse on left half of array
		// Recursion subproblem f(i, m-1)
		return dfs(nums, target, i, m-1)
	} else {
		// Found the target element, return its index
		return m
	}
}

/* Binary search */
func binarySearch(nums []int, target int) int {
	n := len(nums)
	return dfs(nums, target, 0, n-1)
}
